--jvm tries to find Java executable system-wide.

Akirathan opened the pull request:
Fixes #11274

### Pull Request Description

Fixes `--jvm` option, given to the native image. This was failing on my machine, because when given `--jvm` option, the runner was trying to find the `java` executable from the distribution manager's runtime (on my system located in `~/.local/share/enso/runtime`) and it used the first runtime found. But the first runtime on my system is JDK 17.

The `--jvm` option now tries to:
- Find a JDK from the distribution manager that has the same version as the JDK used for building the engine.
  - If there is not an exact version match, it tries to find a runtime from distribution manager that is *newer*.
  - If none, fallback to system-wide search
- System-wide search tries to find `java` from `$JAVA_HOME` and from `$PATH`. But this is just a fallback.

### Important Notes

- Added test to Engine CI jobs that pass `--jvm` argument to a native image of engine-runner
  - ea3af5ffbca6dc439690142251426b4c167bca98
- `runtime-version-manager` sbt project migrated to a JPMS module
- `engine-runner` now depends on `runtime-version-manager`.
- Removed unnecessary stuff in `runtime-version-manager` dealing with outdated `gu` Graal Updater utility.
- Extracted [GraalVersionManager](https://github.com/enso-org/enso/blob/1455b025cb8e5141409ea96b75577e8884265a70/lib/scala/runtime-version-manager/src/main/java/org/enso/runtimeversionmanager/components/GraalVersionManager.java) from [RuntimeVersionManager](https://github.com/enso-org/enso/blob/d2e8994700fd36228f7873f040668013e643190a/lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/components/RuntimeVersionManager.scala)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
--jvm tries to find Java executable system-wide.

JaroslavTulach reviewed (commented):
> The --jvm option now tries to find java executable from JAVA_HOME and from PATH. If that is not found, the distribution manager's runtimes are used as a fallback.

I'd say that this is the **wrong order**. The target audience of Enso users is unlikely to have proper JVM installed in the system. Searching for a system JVM is good for us, developers, but not for our target audience.

> trying to find the java executable from the distribution manager's runtime (on my system located in ~/.local/share/enso/runtime) and it used the first runtime found. But the first runtime on my system is JDK 17.

It is good we have such a failing example! What is the layout of your Enso JVM runtimes? That is an important information for your #11274 bug report.

Anyway, the `enso` launcher should:
- search _managed runtimes_ (in `~/.local/share/enso/runtime`) first
- ignore the outdated ones - e.g. having JVM spec version lower than `21` (right now)
- select the exact match, if found
- select acceptable match (e.g. newer JDK), if found - probably with some warning printed
- if no match found among _managed runtimes_, fallback to system JVM - with some warning printed, if the JDK isn't exactly the one we expect

Would the above steps work for your system setup, Pavel?

Akirathan commented:
> Searching for a system JVM is good for us, developers, but not for our target audience.

That is a good point.

> It is good we have such a failing example! What is the layout of your Enso JVM runtimes? That is an important information for your https://github.com/enso-org/enso/issues/11274 bug report.

```
$ ls -l ~/.local/share/enso/runtime
drwxrwxr-x 1 pavel pavel 612 Nov  1  2023 graalvm-ce-java17.0.7-23.0.0
drwxrwxr-x 1 pavel pavel 114 Nov 28  2023 graalvm-ce-java21.0.1-23.1.0
drwxrwxr-x 1 pavel pavel 114 Feb 29  2024 graalvm-ce-java21.0.2-23.1.2
drwxrwxr-x 1 pavel pavel 114 Apr 16  2024 graalvm-ce-java21.0.2-24.0.0
```

> Would the above steps work for your system setup, Pavel?

Certainly. Sounds like a good plan. I will see what I can do.

JaroslavTulach commented:
> ```
> $ ls -l ~/.local/share/enso/runtime
> drwxrwxr-x 1 pavel pavel 612 Nov  1  2023 graalvm-ce-java17.0.7-23.0.0
> drwxrwxr-x 1 pavel pavel 114 Nov 28  2023 graalvm-ce-java21.0.1-23.1.0
> drwxrwxr-x 1 pavel pavel 114 Feb 29  2024 graalvm-ce-java21.0.2-23.1.2
> drwxrwxr-x 1 pavel pavel 114 Apr 16  2024 graalvm-ce-java21.0.2-24.0.0
> ```

It would be great, if we selected the same JDK as used for the build (however that's `graalvm-community-openjdk-21.0.2+13.1` and I don't see it) and then try any JDK21 (probably the highest - e.g. `graalvm-ce-java21.0.2-24.0.0`) maybe with a warning...

> Sounds like a good plan. I will see what I can do.

[DistributionManager](https://github.com/enso-org/enso/blame/867c77d5cc91056d0fe9987906b54af6696fa13b/lib/scala/distribution-manager/src/main/scala/org/enso/distribution/DistributionManager.scala#L134) shall do most of the selection logic. CCing @radeusgd to give us an _expert guidance_.

Commit:
Remove outdated GraalVMComponentUpdater and related functionality from runtime-version-manager

Commit:
runtime-version-manager does not depend on akka-http

Commit:
Extract GraalVersionManager from RuntimeVersionManager

Commit:
engine-runner depends on runtime-version-manager

Commit:
JavaFinder uses GraalVersionManager to find the java executable

Commit:
Generate javaVersion into version-output

radeusgd commented on engine/runner/src/main/java/org/enso/runner/JavaFinder.java:
```diff
@@ -0,0 +1,64 @@
+final class JavaFinder {
+  private JavaFinder() {}
+
+  /**
+   * Tries to find {@code java} executable on the system. If a system-wide JDK is not found, tries
+   * to find it in the {@link DistributionManager distribution} runtimes.
+   *
+   * @return null if cannot be found. Otherwise, returns the absolute path to the executable, or
+   *     simply {@code java} if it is on the {@code PATH}.
+   */
+  static String findJavaExecutable() {
+    var javaHome = System.getenv("JAVA_HOME");
+    if (javaHome != null) {
+      var java = new File(javaHome, "bin/java").getAbsoluteFile();
+      if (java.exists()) {
+        return java.getAbsolutePath();
+      }
+    }
+    if (isJavaOnPath()) {
+      return "java";
+    }
+    var javaInRuntime = findJavaExecutableInDistributionRuntimes();
+    if (javaInRuntime != null) {
+      return javaInRuntime.toAbsolutePath().toString();
+    }
+    return null;
+  }
```
I think the precedence should actually look into distribution _first_ and only afterwards use system JVM.

The whole purpose of `ensoup` was to manage JVMs related to the engine versions, so that the engine is run with the right compatible JVM and not a random version that may be incompatible with it (I guess it used to be more important before Truffle 'got unchained').

I think ideally we should first to run the JVM version _required by the engine_ if it's found in the distribution. Then we can fallback to using the logic of searching `JAVA_HOME` and `Path`.

Ideally I'd also check the JVM version and log a warning if the version required by engine does not match the version that has been found.

radeusgd replied:
> JVMs related to the engine versions

This version is stored in the engine's manifest - see for example `built-distribution/enso-engine-0.0.0-dev-windows-amd64/enso-0.0.0-dev/manifest.yaml`.

It contains fields: `graal-vm-version` and `graal-java-version`.
You can see `lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/components/Manifest.scala` for the logic of loading the manifest and extracting the `GraalVMVersion` from it. 

We have a `GraalCEReleaseProvider` which translates this `GraalVMVersion` into a strategy for downloading it from GitHub - but you probably don't need this.

The function `graalDirectoryForVersion` in `RuntimeVersionManager` takes `GraalVMVersion` and gives you back the directory name under which this JVM version will be sitting in the `$ENSO_DATA_DIRECTORY/runtime` (installed by `ensoup`).

I guess you could factor out the `Manifest` and `graalDirectoryForVersion` utilities to a common package and you could use it to figure out the JVM version tied to the engine.

----

Btw. the manifest also contains a `jvm-options` field. I'd be very happy if we could keep the _(sometimes changing)_ options in a single place.

radeusgd replied:
I guess if you want to check if the JVM version is correct you need to call `java --version` and somehow parse its output and compare with `graal-java-version`.

e.g. I get
```
openjdk 21.0.2 2024-01-16
OpenJDK Runtime Environment GraalVM CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30)
OpenJDK 64-Bit Server VM GraalVM CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30, mixed mode, sharing)
```

I guess I want to extract the second word from the first line. At least for now that seems to work. But we can update the logic whenever the version format of expected JVM changes - unexpected format just means it's not the JVM we expect :)

JaroslavTulach replied:
Possibly better way than parsing output of `java -version` is to run a small program on the JVM:
```java
public static void main(String... args) {
  System.out.prinltn(System.getProperty("java.vm.version"));
  // etc.
}
```
and just read the output line by line without the need to parse words.

radeusgd replied:
Cool - that indeed sounds better, I didn't think of that

Commit:
GraalVMVersion is comparable

Commit:
Add some tests to SemVer parsing

Commit:
JavaFinder tries to find newer runtime than the one used for building

Commit:
Fix version-output versions mismatch

Akirathan commented:
As of 586fc73aa3e6a8706cbc2e4353a2286ff2c169a6:
- I have extracted [GraalVersionManager](https://github.com/enso-org/enso/blob/1455b025cb8e5141409ea96b75577e8884265a70/lib/scala/runtime-version-manager/src/main/java/org/enso/runtimeversionmanager/components/GraalVersionManager.java) from [RuntimeVersionManager](https://github.com/enso-org/enso/blob/d2e8994700fd36228f7873f040668013e643190a/lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/components/RuntimeVersionManager.scala)
  - Among other things, `GraalVersionManager` is able to list, and sort, all the installed runtimes in the distribution (in my case, it is inside `~/.local/share/enso/runtime`)
- `engine-runner` depends on `runtime-version-manager`
  - For that, I had to convert `runtime-version-manager` project to JPMS module.
- Removed unnecessary stuff in `runtime-version-manager` dealing with outdated `gu` Graal Updater utility.
- JavaFinder knows the exact version that was used for building - https://github.com/enso-org/enso/blob/1455b025cb8e5141409ea96b75577e8884265a70/engine/runner/src/main/java/org/enso/runner/JavaFinder.java#L58
  - And tries to find the runtime from the distribution with the exact version, or newer version

Only when there is no newer runtime found in the managed distribution runtimes, `JAVA_HOME` and system `java` is tried.

Commit:
Update docs

Commit:
runtimeSearchPath does not have to be a directory

Commit:
Fix some build failures in tests

radeusgd commented on lib/scala/runtime-version-manager/src/main/java/org/enso/runtimeversionmanager/components/GraalVersionManager.java:
```diff
@@ -0,0 +1,137 @@
+            "The runtime "
+                + version
+                + "is already installed, but cannot be "
+                + "loaded du to "
```
```suggestion
                + "loaded due to "
```

radeusgd reviewed (approved):
Looks great!

Commit:
docs

JaroslavTulach commented on build.sbt:
```diff
@@ -4328,15 +4332,34 @@ lazy val `connected-lock-manager-server` = project
+      "org.slf4j"          % "slf4j-api"        % slf4jVersion
+    ),
+    Compile / internalModuleDependencies := Seq(
+      (`cli` / Compile / exportedModule).value,
```
`cli`? Is that really needed?

Akirathan replied:
This dependency has been there for ages: https://github.com/enso-org/enso/blob/b89db13046a174b815afb95e90143de5698b80d2/build.sbt#L4376. And yes, it is needed, because `runtime-version-manager` uses types from `cli`.

JaroslavTulach commented on build.sbt:
```diff
@@ -4328,15 +4332,34 @@ lazy val `connected-lock-manager-server` = project
+    Compile / internalModuleDependencies := Seq(
+      (`cli` / Compile / exportedModule).value,
+      (`distribution-manager` / Compile / exportedModule).value,
+      (`downloader` / Compile / exportedModule).value,
```
`downloader`? Right now all downloads should be disabled. Necessary?

Akirathan replied:
`runtime-version-manager` contains [GithubAPI](https://github.com/enso-org/enso/blob/f3bc88b91e5f694467935a417552d65daf92bb86/lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/releases/github/GithubAPI.scala) that uses stuff from `downloader`. But it is not actually called. Strictly speaking, it is not necessary, but I don't want to try to split the code right now.

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/JavaFinder.java:
```diff
@@ -0,0 +1,91 @@
+    logger.warn("No appropriate JDK found in the distribution runtimes. Trying system-wide JDK.");
+    var javaHome = System.getenv("JAVA_HOME");
+    if (javaHome != null) {
+      var java = new File(javaHome, "bin/java").getAbsoluteFile();
```
This is not Windows `\` friendly. Try: `new File(new File(javaHome, "bin"), "java")` - btw. where is the `.exe` extension for Windows!?

Akirathan replied:
Made Windows friendly in fd52bbf63a3ea7a0c05d66243bbf18dfe8b210a8

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/JavaFinder.java:
```diff
@@ -0,0 +1,91 @@
+      }
+    }
+    if (isJavaOnPath()) {
+      return "java";
```
Does that work on Windows?

Akirathan replied:
I don't expect it to work on windows. This is a fallback that is currently only usable for devs and not anyone else. For all the other use-cases, the java exe should be found in the installed distribution runtimes.

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/java/module-info.java:
```diff
@@ -0,0 +1,22 @@
+  exports org.enso.runtimeversionmanager;
+  exports org.enso.runtimeversionmanager.cli;
+  exports org.enso.runtimeversionmanager.components;
+  exports org.enso.runtimeversionmanager.runner;
```
That's a lot of exported packages. Shouldn't the API be cleaned and put into a single package?

Akirathan replied:
It should. Like all the other Scala-only projects that are now historical remnants. I suggest revisiting this once we take a look into `launcher` (`ensoup`) and related projects.

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/java/org/enso/runtimeversionmanager/components/GraalVersionManager.java:
```diff
@@ -0,0 +1,140 @@
+/**
+ * Utility class that finds installed managed runtimes (Graal JDK) from {@link DistributionManager}.
+ */
+public class GraalVersionManager {
```
`
```suggestion
public final class GraalVersionManager {
```

JaroslavTulach reviewed (approved):
Overall it looks very nice and clean. Some doubts:
- does it work on Windows?
- possibly too many exports from the new `module-info`

Commit:
Merge branch 'develop' into wip/akirathan/11274-jvm-cmdlineopt-ni

Commit:
Make JavaFinder Windows friendly

Commit:
GraalVersionManager is final

Commit:
Add test for --jvm arg

Commit:
Log found JDK

Commit:
Add timeout to waiting for subprocess

Commit:
Update licenses - akka-http was removed from launcher

Commit:
When testing --jvm in CI, don't override ENSO_DATA_DIRECTORY env var

Commit:
Run --jvm test only on Linux

Akirathan commented on build/build/src/engine/context.rs:
```diff
@@ -686,7 +687,25 @@ pub async fn runner_sanity_test(
+
+        // The following test does not actually run anything, it just checks if the engine
+        // can accept `--jvm` argument and evaluates something.
+        if TARGET_OS != OS::Windows {
```
When the test runs on the CI, there is nothing inside `ENSO_DATA_DIRECTORY`. The `--jvm` argument first tries to find any appropriate JDK in the `ENSO_DATA_DIRECTORY/runtime` directory, but on the CI, it is empty. The fallback is to search in `JAVA_HOME` env var or on `PATH`, but that works only on Linux. Let's just keep this sanity check on Linux for now.

Akirathan replied:
Trying to run the tests also on Windows again in https://github.com/enso-org/enso/pull/11500/commits/fb25d57ed8cad3f563489440353444eaec4d6706

hubertp replied:
So that didn't work?

Commit:
On Windows, java.exe is tried to be found on PATH

Commit:
Revert "Run --jvm test only on Linux"

This reverts commit f14333d1f2ded28d83b18a445c1f4faca5d01cea.

Commit:
Reapply "Run --jvm test only on Linux"

This reverts commit fb25d57ed8cad3f563489440353444eaec4d6706.

Akirathan commented:
Windows CI does not have `java.exe` on PATH - https://github.com/enso-org/enso/actions/runs/11840014687/job/32992849858?pr=11500#step:7:2539 . Removing this test from Windows again in 049b991eaa4c4a699c43a9a73ac2f3fa43de5219

<blockquote><img src="https://opengraph.githubassets.com/5f37b72feef865df45390f375907e2b75b231ebad70dec5202e53c5766227aee/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">--jvm tries to find Java executable system-wide. · enso-org/enso@fb25d57</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - --jvm tries to find Java executable system-wide. · enso-org/enso@fb25d57</div></blockquote>

radeusgd commented:
> Windows CI does not have `java.exe` on PATH - [enso-org/enso/actions/runs/11840014687/job/32992849858?pr=11500#step:7:2539](https://github.com/enso-org/enso/actions/runs/11840014687/job/32992849858?pr=11500#step:7:2539) . Removing this test from Windows again in [049b991](https://github.com/enso-org/enso/commit/049b991eaa4c4a699c43a9a73ac2f3fa43de5219)
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/d68cb9fd06df5b9e3a1b8bf2af8089dbcc50e4c1dfa75723fbbc20d631e257ec/68747470733a2f2f6f70656e67726170682e6769746875626173736574732e636f6d2f356633376237326665656638363564663435333930663337353930376532623735623233316562616437306465633532303265353363353736363232376165652f656e736f2d6f72672f656e736f"><img alt="" height="14" src="https://camo.githubusercontent.com/c7475aac04b72ccdbdab6d5f8abf505c824fdb625ef049acf2c773106c631bff/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f66617669636f6e732f66617669636f6e2e737667"> GitHub**[--jvm tries to find Java executable system-wide. · enso-org/enso@fb25d57](https://github.com/enso-org/enso)**Enso Analytics is a self-service data prep and analysis platform designed for data teams. - --jvm tries to find Java executable system-wide. · [fb25d57](https://github.com/enso-org/enso/commit/fb25d57)

What about `JAVA_HOME/bin`?

hubertp commented on engine/runner/src/main/java/org/enso/runner/JavaFinder.java:
```diff
@@ -0,0 +1,110 @@
+      logger.warn("Falling back to java on PATH: {}", javaExe);
+      return javaExe;
+    }
+    logger.warn("No JDK found on PATH. Cannot start the runtime.");
```
Shouldn't you be checking `ENSO_JVM_PATH` as well?

hubertp commented on engine/runner/src/main/java/org/enso/runner/JavaFinder.java:
```diff
@@ -0,0 +1,110 @@
+    return null;
+  }
+
+  private static boolean isOnWindows() {
```
Could just use `OS.isWindows()` instead? It's already a dependency, right?

hubertp commented on engine/runner/src/main/java/org/enso/runner/JavaFinder.java:
```diff
@@ -0,0 +1,110 @@
+    try {
+      ProcessBuilder processBuilder;
+      if (isOnWindows()) {
+        processBuilder = new ProcessBuilder("java.exe", "-h");
```
nit: 
```suggestion
        processBuilder = new ProcessBuilder("java.exe", "-v");
```

hubertp commented on engine/runner/src/main/java/org/enso/runner/JavaFinder.java:
```diff
@@ -0,0 +1,110 @@
+      if (isOnWindows()) {
+        processBuilder = new ProcessBuilder("java.exe", "-h");
+      } else {
+        processBuilder = new ProcessBuilder("java", "-h");
```
```suggestion
        processBuilder = new ProcessBuilder("java", "-v");
```

hubertp reviewed (approved):
I'm a bit surprised this fails Windows PR check. What's the reason?

JaroslavTulach commented:
Congratulation to getting thru CI checks.
